graphbolt/src/feature_cache.h

```diff
@@ -0,0 +1,202 @@
+
+  T* Push(const T& x) {
+    auto insert_ptr = &data_[PostIncrement(tail_)];
+    *insert_ptr = x;
+    return insert_ptr;
+  }
+
+  T Pop() { return data_[PostIncrement(head_)]; }
+
+  T& Front() const { return data_[head_]; }
+
+  bool IsFull() const {
+    const auto diff = tail_ + 1 - head_;
+    return diff == 0 || diff == capacity_;
+  }
+
+  friend std::ostream& operator<<(
+      std::ostream& os, const CircularQueue& queue) {
+    for (auto i = queue.head_; i != queue.tail_; queue.PostIncrement(i)) {
+      os << queue.data_[i] << ", ";
+    }
+    return os << "\n";
+  }
+
+  bool IsEmpty() const { return tail_ == head_; }
+
+  int64_t Capacity() const { return capacity_ - 1; }
+
+ private:
+  int64_t PostIncrement(int64_t& i) const {
+    const auto ret = i++;
+    if (i >= capacity_) i -= capacity_;
+    return ret;
+  }
+
+  int64_t tail_;
+  int64_t head_;
+  int64_t capacity_;
+  std::unique_ptr<T[]> data_;
+};
+
+struct CacheKey {
+  int64_t freq_ : 3;
+  int64_t key_ : 61;
+  int64_t position_;
+
+  CacheKey(int64_t key, int64_t position)
+      : freq_(0), key_(key), position_(position) {
+    static_assert(sizeof(CacheKey) == 2 * sizeof(int64_t));
+  }
+
+  CacheKey() = default;
+
+  void Increment() { freq_ = std::min(3, static_cast<int>(freq_ + 1)); }
+
+  void Decrement() { freq_ = std::max(0, static_cast<int>(freq_ - 1)); }
+
+  friend std::ostream& operator<<(std::ostream& os, const CacheKey& key_ref) {
+    return os << '(' << key_ref.key_ << ", " << key_ref.freq_ << ", "
+              << key_ref.position_ << ")";
+  }
+};
+
+class S3FifoCachePolicy : public torch::CustomClassHolder {
+ public:
+  /**
+   * @brief Constructor for the S3FifoCachePolicy class.
+   *
+   * @param capacity The capacity of the cache in terms of # elements.
+   */
+  S3FifoCachePolicy(int64_t capacity);
+
+  S3FifoCachePolicy() = default;
+
+  std::tuple<torch::Tensor, torch::Tensor, torch::Tensor> Query(
+      torch::Tensor keys);
+
+  torch::Tensor Replace(torch::Tensor keys);
+
+  static c10::intrusive_ptr<S3FifoCachePolicy> Create(int64_t capacity);
+
+  friend std::ostream& operator<<(
+      std::ostream& os, const S3FifoCachePolicy& m) {
+    return os << "S_: " << m.S_ << "\n"
+              << "M_: " << m.M_ << "\n"
+              << "position_q_: " << m.position_q_ << "\n"
+              << "ghost_q_time_: " << m.ghost_q_time_ << "\n"
+              << "capacity_: " << m.capacity_ << "\n";
+  }
+
+ private:
+  int64_t EvictM() {
+    for (;;) {
+      auto evicted = M_.Pop();
+      auto it = position_map_.find(evicted.key_);
+      if (evicted.freq_ > 0) {
+        evicted.Decrement();
+        it->second = M_.Push(evicted);
+      } else {
+        position_map_.erase(it);
+        return evicted.position_;
+      }
+    }
+  }
+
+  int64_t EvictS() {
+    auto evicted = S_.Pop();
+    const auto is_M_full = M_.IsFull();
+    auto it = position_map_.find(evicted.key_);
+    if (evicted.freq_ > 0 || !is_M_full) {
```

Comment:
Swap it: avoid if not as much as possible.
if (evicted.freq_ == 0 && is_M_full) {